addons/tls13-middlebox/bootstrap.js

```diff
@@ -0,0 +1,224 @@
+
+const VERSION_MAX_PREF = "security.tls.version.max";
+const FALLBACK_LIMIT_PREF = "security.tls.version.fallback-limit";
+
+let readwrite_prefs = new Preferences({defaultBranch: true});
+
+// all combination of configurations we care about.
+// for is_tls13 == true we need a server that supports TLS 1.3
+// for is_tls13 == false we need a server that DOES NOT support TLS 1.3
+let configurations = [
+    {max_version: 4, fallback_limit: 4, is_tls13: true, website: "enabled.tls13.com"},
+    {max_version: 4, fallback_limit: 4, is_tls13: false, website: "disabled.tls13.com"},
+    {max_version: 4, fallback_limit: 3, is_tls13: true, website: "tls13.crypto.mozilla.org"},
+    {max_version: 4, fallback_limit: 3, is_tls13: false, website: "control.tls12.com"},
+    {max_version: 3, fallback_limit: 3, is_tls13: true, website: "www.allizom.org"},
+    {max_version: 3, fallback_limit: 3, is_tls13: false, website: "short.tls13.com"}
+];
+
+function getFieldValue(obj, name) {
+	try {
+		return obj[name];
+	} catch (ex) {
+		return undefined;
+	}
+}
+
+function getInfo(xhr) {
+    let result = {};
+
+    try {
+        let channel = xhr.channel;
+
+        // this is the most important value based on which we can find out the problem
+        channel.QueryInterface(Ci.nsIRequest);
+        result.status = getFieldValue(channel, 'status');
+
+        let securityInfo = getFieldValue(channel, 'securityInfo');
+
+        if (securityInfo instanceof Ci.nsITransportSecurityInfo) {
+            securityInfo.QueryInterface(Ci.nsITransportSecurityInfo);
+
+            result.securityState = getFieldValue(securityInfo, 'securityState');
+
+            // Error message on connection failure. I am not sure if we can get this error message using status code.
+            // It is safer to collect this information as well.
+            result.errorMessage = getFieldValue(securityInfo, 'errorMessage');
+        }
+
+        if (securityInfo instanceof Ci.nsISSLStatusProvider) {
+            securityInfo.QueryInterface(Ci.nsISSLStatusProvider);
+            let sslStatus = securityInfo.SSLStatus;
+
+            if (sslStatus) {
+                sslStatus.QueryInterface(Ci.nsISSLStatus);
+                let serverCert = sslStatus.serverCert;
+
+                result.certChain = [];
+
+                // extracting the certificate chain including the root CA.
+                // if isBuiltInRoot == false, it means there is middlebox on the way
+                let cert = serverCert;
+                while (cert) {
+                	result.certChain.push({
+                		certType: getFieldValue(cert, 'certType'),
+                		isBuiltInRoot: getFieldValue(cert, 'isBuiltInRoot'),
+                		isSelfSigned: getFieldValue(cert, 'isSelfSigned'),
+                		keyUsages: getFieldValue(cert, 'keyUsages')
+                	});
+
+                	cert = getFieldValue(cert, 'issuer');
+                }
```

Comment:
I would just record:

(1) the hash of the leaf cert
(2) whether the final cert is isBuiltInRoot

You will need to test this with a non-builtin-a